File: database/src/routes/content.ts
import { Router, type Request, type Response } from "express";
import { get_movie_by_id } from "../movieService.js";
import { get_series_by_id } from "../seriesService.js";
import { get_home_rails, get_content_by_genre } from "../libraryService.js";
import { Genre } from "../models/Genre.js";
import { optional_auth } from "../middleware/auth.js";

const router = Router();

// GET /api/content/home — Netflix-style home page rails
// Uses optional auth so Continue Watching rail appears for logged-in users
router.get("/home", optional_auth, async (req: Request, res: Response) => {
    try {
        const user_id = req.user?.user_id;
        const data = await get_home_rails(user_id);
        res.json(data);
    } catch (error) {
        res.status(500).json({ error: (error as Error).message });
    }
});

// GET /api/content/genres — list all genres
router.get("/genres", async (_req: Request, res: Response) => {
    try {
        const genres = await Genre.find().sort({ name: 1 }).lean();
        res.json(genres);
    } catch (error) {
        res.status(500).json({ error: (error as Error).message });
    }
});

// GET /api/content/browse/:genre_id — browse content by genre
router.get("/browse/:genre_id", async (req: Request, res: Response) => {
    try {
        const genre_id = Number(req.params.genre_id);
        const page = req.query.page ? Number(req.query.page) : 1;
        const limit = req.query.limit ? Number(req.query.limit) : 20;
        const data = await get_content_by_genre(genre_id, page, limit);
        res.json(data);
    } catch (error) {
        res.status(500).json({ error: (error as Error).message });
    }
});

// GET /api/content/movie/:id — get movie details
router.get("/movie/:id", async (req: Request, res: Response) => {
    try {
        const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
        const movie = await get_movie_by_id(parseInt(id));
        res.json(movie);
    } catch (error) {
        res.status(404).json({ error: (error as Error).message });
    }
});

// GET /api/content/series/:id — get series details with seasons/episodes
router.get("/series/:id", async (req: Request, res: Response) => {
    try {
        const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
        const series = await get_series_by_id(parseInt(id));
        res.json(series);
    } catch (error) {
        res.status(404).json({ error: (error as Error).message });
    }
});

export default router;

File: fastapi-backend/app/routers/content.py
"""
Content router — mirrors database/src/routes/content.ts.
GET /home, GET /genres, GET /browse/:genre_id, GET /movie/:id, GET /series/:id, GET /search
"""

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from app.core.deps import get_optional_current_user
from app.database import get_database
from app.services.library_service import get_content_by_genre, get_home_rails
from app.services.movie_service import get_movie_by_id
from app.services.series_service import get_series_by_id
from app.utils import sanitize


router = APIRouter(prefix="/api/content", tags=["content"])


@router.get("/home")
async def home(
    limit: int = Query(10, ge=1, le=50),
    current_user: Optional[dict] = Depends(get_optional_current_user),
):
    user_id = current_user["_id"] if current_user else None
    return await get_home_rails(user_id, rail_limit=limit)


@router.get("/genres")
async def list_genres():
    db = get_database()
    cursor = db.genres.find({"is_hidden": {"$ne": True}}).sort("name", 1)
    genres = []
    async for doc in cursor:
        genres.append(sanitize(doc))
    return genres


@router.get("/browse/{genre_id}")
async def browse_genre(
    genre_id: int,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
):
    db = get_database()
    genre = await db.genres.find_one({"genre_id": genre_id})
    if genre and genre.get("is_hidden"):
        raise HTTPException(status_code=404, detail="Genre not found")
    return await get_content_by_genre(genre_id, page, limit)


@router.get("/search")
async def search_content(
    q: str = Query(..., min_length=1),
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
):
    skip = (page - 1) * limit
    db = get_database()
    escaped = q.replace(".", r"\.")

    # Search both movies and series
    movie_cursor = db.movies.find(
        {"title": {"$regex": escaped, "$options": "i"}, "is_hidden": {"$ne": True}},
        {"tmdb_id": 1, "title": 1, "poster_path": 1, "backdrop_path": 1, "vote_average": 1, "release_date": 1, "overview": 1},
    ).limit(limit + skip)
    series_cursor = db.series.find(
        {"name": {"$regex": escaped, "$options": "i"}, "is_hidden": {"$ne": True}},
        {"tmdb_id": 1, "name": 1, "poster_path": 1, "backdrop_path": 1, "vote_average": 1, "first_air_date": 1, "overview": 1},
    ).limit(limit + skip)

    movies = [dict(sanitize(d), content_type="movie") async for d in movie_cursor]
    series = [dict(sanitize(d), content_type="series") async for d in series_cursor]

    # Merge and paginate
    all_results = movies + series
    return {
        "results": all_results[skip : skip + limit],
        "page": page,
        "limit": limit,
        "total": len(all_results),
    }


@router.get("/movie/{movie_id}")
async def movie_detail(movie_id: int):
    try:
        return await get_movie_by_id(movie_id)
    except Exception as exc:
        raise HTTPException(status_code=404, detail=str(exc))


@router.get("/series/{series_id}")
async def series_detail(series_id: int):
    try:
        return await get_series_by_id(series_id)
    except Exception as exc:
        raise HTTPException(status_code=404, detail=str(exc))
